middle_end/flambda/types/env/aliases.ml

```diff
@@ -289,127 +418,195 @@ let add_alias_between_canonical_elements t ~canonical_element ~to_be_demoted =
-    let { canonical_element; to_be_demoted; } =
-      choose_canonical_element_to_be_demoted t ~canonical_element1
-        ~canonical_element2
-    in
-    let t =
-      add_alias_between_canonical_elements t ~canonical_element
-        ~to_be_demoted
-    in
-    { t;
-      canonical_element;
-      (* CR mshinwell: [alias_of] is not a good name. *)
-      alias_of = to_be_demoted;
-    }
+  | Is_canonical canonical_element1, Is_canonical canonical_element2
   | Alias_of_canonical
-        { element = element1; canonical_element = canonical_element1; },
+        { element = _; canonical_element = canonical_element1; },
       Is_canonical canonical_element2
-  | Is_canonical canonical_element2,
+  | Is_canonical canonical_element1,
       Alias_of_canonical
-        { element = element1; canonical_element = canonical_element1; } ->
-    let { canonical_element; to_be_demoted; } =
-      choose_canonical_element_to_be_demoted t ~canonical_element1
-        ~canonical_element2
-    in
-    let alias_of =
-      if Simple.equal to_be_demoted canonical_element1 then element1
-      else canonical_element2
-    in
-    let t =
-      add_alias_between_canonical_elements t ~canonical_element
-        ~to_be_demoted
-    in
-    { t;
-      canonical_element;
-      alias_of;
-    }
+        { element = _; canonical_element = canonical_element2; }
   | Alias_of_canonical
-        { element = element1; canonical_element = canonical_element1; },
+        { element = _; canonical_element = canonical_element1; },
       Alias_of_canonical
-        { element = element2; canonical_element = canonical_element2; }
+        { element = _; canonical_element = canonical_element2; }
       ->
-    let { canonical_element; to_be_demoted; } =
-      choose_canonical_element_to_be_demoted t ~canonical_element1
-        ~canonical_element2
-    in
-    let alias_of =
-      if Simple.equal to_be_demoted canonical_element1 then element1
-      else element2
-    in
-    let t =
-      add_alias_between_canonical_elements t ~canonical_element
-        ~to_be_demoted
-    in
-    { t;
-      canonical_element;
-      alias_of;
-    }
+    if Simple.equal canonical_element1 canonical_element2
+    then
+      let canonical_element = canonical_element1 in
+      (* According to the contract for [add], [alias_of_demoted_element] must
+         not be canonical. Usually this is fine, but what if [element1] or
+         [element2] is *itself* canonical? This is true iff that element is
+         equal to [canonical_element1]. In that case, we can safely pick the
+         other element. (They cannot both be canonical because then they'd both
+         be equal to [canonical_element1] and we assume that [element1] and
+         [element2] are different.) *)
+      (* CR lmaurer: We should just bail out in this case; since [element1] and
+         [element2] have the same canonical, they're already aliases, so
```

Comment:
I prefer not to bail out in this case. We are making some assumption on how [Typing_env.add_equation] and there's no guarantees for it to behave like this in the future. This could introduce a subtile and hard to debug bug later on.

Reply:
I had in mind something like adding a constructor to `add_result` saying that they were already aliases. Maybe worth adding that to the comment?

Reply:
Added to the comment.

Reply:
I'll reformulate. Is this behavior (bailing out) the one we want in all cases, even if aliases were to be used with something else than [Typing_env.add_equation]? If that's the case then we should remove the misleading reference to [Typing_env.add_equation] from the comment.

Reply:
I think so, yes. Currently we're basically lying (since we didn't actually change anything) in such a way as to make sure [Typing_env.add_equation] doesn't break. Improving the API would reduce the coupling. I've revised the comments so they're no longer vague about what "bail out" means.